`Installing mysql2 0.3.19 with native extensions` Hangs forever

on ruby 2.2.2 with rbenv 0.4.0
both on a Mac OS X 10.9.5 
and Ubuntu 12.04 (fresh vagrant vm spun up with test-kitchen)
`bundle install` (and `gem install mysql2 -v 0.3.19`)  hangs forever installing 
the
`/opt/rbenv/versions/2.2.2/lib/ruby/gems/2.2.0/gems/mysql2-0.3.19/ext/mysql2/mkmf.log` just fills with a bunch of repeats of this
```text

"gcc -o conftest -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0/x86_64-linux -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0/ruby/backward -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0 -I. -I/usr/local/include -I/opt/rbenv/versions/2.2.2/include     -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat conftest.c  -L/usr/local/lib -Wl,-R/usr/local/lib -L. -L/opt/rbenv/versions/2.2.2/lib -Wl,-R/opt/rbenv/versions/2.2.2/lib -L. -L/opt/rbenv/versions/2.2.2/lib  -fstack-protector -rdynamic -Wl,-export-dynamic    -lm  -Wl,-R/opt/rbenv/versions/2.2.2/lib -L/opt/rbenv/versions/2.2.2/lib -lruby-static -lmysqlclient -lm  -lpthread -lrt -ldl -lcrypt -lm   -lc"
conftest.c: In function ‘t’:
conftest.c:13:57: error: ‘mysql_query’ undeclared (first use in this function)
conftest.c:13:57: note: each undeclared identifier is reported only once for each function it appears in
conftest.c:13:32: warning: variable ‘p’ set but not used [-Wunused-but-set-variable]
checked program was:
/* begin */
 1: #include "ruby.h"
 2:
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10:
11:   return 0;
12: }
13: int t(void) { void ((*volatile p)()); p = (void ((*)()))mysql_query; return 0; }
/* end */

"gcc -o conftest -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0/x86_64-linux -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0/ruby/backward -I/opt/rbenv/versions/2.2.2/include/ruby-2.2.0 -I. -I/usr/local/include -I/opt/rbenv/versions/2.2.2/include     -O3 -fno-fast-math -ggdb3 -Wall -Wextra -Wno-unused-parameter -Wno-parentheses -Wno-long-long -Wno-missing-field-initializers -Wunused-variable -Wpointer-arith -Wwrite-strings -Wdeclaration-after-statement -Wimplicit-function-declaration -Wdeprecated-declarations -Wno-packed-bitfield-compat conftest.c  -L/usr/local/lib -Wl,-R/usr/local/lib -L. -L/opt/rbenv/versions/2.2.2/lib -Wl,-R/opt/rbenv/versions/2.2.2/lib -L. -L/opt/rbenv/versions/2.2.2/lib  -fstack-protector -rdynamic -Wl,-export-dynamic    -lm  -Wl,-R/opt/rbenv/versions/2.2.2/lib -L/opt/rbenv/versions/2.2.2/lib -lruby-static -lmysqlclient -lm  -lpthread -lrt -ldl -lcrypt -lm   -lc"
conftest.c: In function ‘t’:
conftest.c:13:1: warning: implicit declaration of function ‘mysql_query’ [-Wimplicit-function-declaration]
/usr/bin/ld: cannot find -lmysqlclient
collect2: ld returned 1 exit status
checked program was:
/* begin */
 1: #include "ruby.h"
 2:
 3: /*top*/
 4: extern int t(void);
 5: int main(int argc, char **argv)
 6: {
 7:   if (argc > 1000000) {
 8:     printf("%p", &t);
 9:   }
10:
11:   return 0;
12: }
13: int t(void) { mysql_query(); return 0; }
/* end */
```

--- a/ext/mysql2/extconf.rb
+++ b/ext/mysql2/extconf.rb
@@ -73,14 +73,21 @@ elsif mc = (with_config('mysql-config') || Dir[GLOB].first)
   rpath_dir = libs
 else
   inc, lib = dir_config('mysql', '/usr/local')
-  libs = ['m', 'z', 'socket', 'nsl', 'mygcc']
-  found = false
-  while not find_library('mysqlclient', 'mysql_query', lib, "#{lib}/mysql") do
-    exit 1 if libs.empty?
-    found ||= have_library(libs.shift)
-  end
+  unless find_library('mysqlclient', 'mysql_query', lib, "#{lib}/mysql")
+    found = false
+    # For some systems and some versions of libmysqlclient, there were extra
+    # libraries needed to link. Try each typical extra library, add it to the
+    # global compile flags, and see if that allows us to link libmysqlclient.
+    warn "-----\nlibmysqlclient is missing. Trying again with extra runtime libraries...\n-----"
 
-  asplode("mysql client") unless found
+    %w{ m z socket nsl mygcc }.each do |extra_lib|
+      if have_library(extra_lib) && find_library('mysqlclient', 'mysql_query', lib, "#{lib}/mysql")
+        found = true
+        break
+      end
+    end
+    asplode('libmysqlclient') unless found
+  end
 
   rpath_dir = lib
 end

--- a/ext/mysql2/extconf.rb
+++ b/ext/mysql2/extconf.rb
@@ -3,12 +3,11 @@ require 'mkmf'
 
 def asplode lib
   if RUBY_PLATFORM =~ /mingw|mswin/
-    abort "-----\n#{lib} is missing. please check your installation of mysql and try again.\n-----"
+    abort "-----\n#{lib} is missing. Check your installation of MySQL or Connector/C, and try again.\n-----"
   elsif RUBY_PLATFORM =~ /darwin/
-    abort "-----\n#{lib} is missing. Try 'brew install mysql', check your installation of mysql and try again.\n-----"
+    abort "-----\n#{lib} is missing. You may need to 'brew install mysql' or 'port install mysql', and try again.\n-----"
   else
-    abort "-----\n#{lib} is missing. Try 'apt-get install libmysqlclient-dev' or
-'yum install mysql-devel', check your installation of mysql and try again.\n-----"
+    abort "-----\n#{lib} is missing. You may need to 'apt-get install libmysqlclient-dev' or 'yum install mysql-devel', and try again.\n-----"
   end
 end
 

